src/main/scala/is/hail/methods/KinshipMatrix.scala

```diff
@@ -0,0 +1,37 @@
+package is.hail.methods
+
+import org.apache.spark.mllib.linalg.Vectors
+import org.apache.spark.mllib.linalg.distributed.{IndexedRow, IndexedRowMatrix}
+import scala.collection.Searching._
+
+/**
+  * Represents a KinshipMatrix, which is a matrix where the number at index (i, j) represents how related the samples at index
+  * i and j in the samples array are.
+  */
+class KinshipMatrix(val matrix: IndexedRowMatrix, val sampleIds: Array[String]) {
+  assert(matrix.numCols().toInt == matrix.numRows().toInt && matrix.numCols().toInt == sampleIds.length)
+
+  /**
+    * Filters list of samples based on predicate, and removes corresponding row and column from the list.
+    * @param pred The predicate that decides whether a sample is kept.
+    */
+  def filterSamples(pred: (String => Boolean)): KinshipMatrix = {
+    val (samplePairsToTake, samplePairsToDrop) = sampleIds.zipWithIndex.partition(pair => pair match {case (sampleName, index) => pred(sampleName)})
```

Comment:
I find samplePairs very confusing in this kinship context, when it's really sampleID and index. How about `sampleWithIndex`? Then using `pred(pair._1)` in place of `pair match {case (sampleName, index) => pred(sampleName)}` will be clearer.